README.rst
==========
fio-buffer
==========

.. image:: https://travis-ci.org/geowurster/fio-buffer.svg?branch=master
    :target: https://travis-ci.org/geowurster/fio-buffer?branch=master

.. image:: https://coveralls.io/repos/geowurster/fio-buffer/badge.svg?branch=master
    :target: https://coveralls.io/r/geowurster/fio-buffer?branch=master

A `Fiona <http://toblerity.org/fiona/manual.html>`_  CLI plugin for buffering geometries in parallel.  Powered by `Shapely <http://toblerity.org/shapely/manual.html#object.buffer>`_.


Usage
=====

.. code-block:: console

    Usage: fio buffer [OPTIONS] INFILE OUTFILE

      Geometries can be dilated with a positive distance, eroded with a negative
      distance, and in some cases cleaned or repaired with a distance of 0.

      Examples
      --------

      Default settings - buffer geometries in the input CRS:

          $ fio buffer in.geojson out.geojson --distance 10

      Dynamically buffer geometries by a distance stored in the field
      'magnitude' and write as GeoJSON:

          $ fio buffer \
              in.shp \
              out.geojson \
              --driver GeoJSON \
              --distance magnitude

      Read geometries from one CRS, buffer in another, and then write to a
      third:

          $ fio buffer in.shp out.shp \
              --distance 10 \
              --buf-crs EPSG:3857 \
              --dst-crs EPSG:32618

      Control cap style, mitre limit, segment resolution, and join style:

          $ fio buffer in.geojson out.geojson \
              --distance 0.1 \
              --res 5 \
              --cap-style flat \
              --join-style mitre \
              --mitre-limit 0.1\

    Options:
      --version                       Show the version and exit.
      -f, --format, --driver NAME     Output driver name.  Derived from the input
                                      datasource if not given.
      --cap-style [flat|round|square]
                                      Where geometries terminate, use this style.
                                      [default: round]
      --join-style [round|mitre|bevel]
                                      Where geometries touch, use this style.
                                      [default: round]
      --res INTEGER                   Resolution of the buffer around each vertex
                                      of the geometry.  [default: 16]
      --mitre-limit FLOAT             When using a mitre join, limit the maximum
                                      length of the join corner according to this
                                      ratio.  [default: 5.0]
      --distance FLOAT|FIELD          Buffer distance or field containing distance
                                      values.  Units match --buf-crs.  When
                                      buffering with a field, feature's with a
                                      null value are unaltered.  [required]
      --src-crs TEXT                  Specify CRS for input data.  Not needed if
                                      set in input file.
      --buf-crs TEXT                  Perform buffer operations in a different
                                      CRS. [default: --src-crs]
      --dst-crs TEXT                  Reproject geometries to a different CRS
                                      before writing.  Must be combined with
                                      --buf-crs. [default: --src-crs]
      --geom-type GEOMTYPE            Output layer's geometry type.  [default:
                                      MultiPolygon]
      --skip-failures                 Skip geometries that fail somewhere in the
                                      processing pipeline.
      --jobs CORES                    Process geometries in parallel across N
                                      cores.  Feature ID's and order are not
                                      preserved if more that 1 cores are used.
                                      [default: 1]
      --help                          Show this message and exit.


Installing
==========

Via pip:

.. code-block:: console

    $ pip install fio-buffer

From source:

.. code-block:: console

    $ git clone https://github.com/geowurster/fio-buffer
    $ cd fio-buffer
    $ python setup.py install


Developing
==========

.. code-block:: console

    $ git clone https://github.com/geowurster/fio-buffer
    $ cd fio-buffer
    $ virtualenv venv
    $ source venv/bin/activate
    $ pip install -e .[dev]
    $ py.test tests --cov fio_buffer --cov-report term-missing


License
=======

See ``LICENSE.txt``.


CHANGES.txt
Changes
=======


0.1.1 (2015-07-21)
------------------

- Initial release


0.1 - (2015-06-08)
------------------

- Test release


tests/test_cli.py
from click.testing import CliRunner
import fiona as fio
import fiona.fio.main
from fiona.transform import transform_geom

from shapely.geometry import mapping
from shapely.geometry import shape

import fio_buffer.core


def test_standard(tmpdir):
    outfile = str(tmpdir.mkdir('out').join("buf-1.geojson"))
    result = CliRunner().invoke(fio_buffer.core.buffer, [
        'tests/data/points.geojson',
        outfile,
        '--distance', '1',
        '--driver', 'GeoJSON'
    ])
    assert result.exit_code == 0
    with fio.open('tests/data/points.geojson') as points, fio.open(outfile) as actual:
        for pnt, buf in zip(points, actual):
            e_coords = mapping(shape(pnt['geometry']).buffer(1))['coordinates'][0]
            a_coords = buf['geometry']['coordinates'][0]
            for e_pair, a_pair in zip(e_coords, a_coords):

                # Multipolygons can creep in and mess up the test so just skip them
                if buf['geometry']['type'] == 'Polygon':
                    e_x, e_y = e_pair
                    a_x, a_y = a_pair
                    assert round(e_x, 3) == round(a_x, 3)
                    assert round(e_y, 3) == round(a_y, 3)


def test_distance_field(tmpdir):
    outfile = str(tmpdir.mkdir('out').join("buf-field.geojson"))
    result = CliRunner().invoke(fio_buffer.core.buffer, [
        'tests/data/points.geojson',
        outfile,
        '--distance', 'distance',
        '--driver', 'GeoJSON'
    ])
    assert result.exit_code == 0
    with fio.open('tests/data/points.geojson') as points, fio.open(outfile) as actual:
        for pnt, buf in zip(points, actual):
            dist = buf['properties']['distance']
            e_coords = mapping(shape(pnt['geometry']).buffer(dist))['coordinates'][0]
            a_coords = buf['geometry']['coordinates'][0]
            for e_pair, a_pair in zip(e_coords, a_coords):
                # Multipolygons can creep in and mess up the test so just skip them
                if buf['geometry']['type'] == 'Polygon':
                    e_x, e_y = e_pair
                    a_x, a_y = a_pair
                    assert round(e_x, 3) == round(a_x, 3)
                    assert round(e_y, 3) == round(a_y, 3)


def test_buf_crs(tmpdir):
    outfile = str(tmpdir.mkdir('out').join("buf-field.geojson"))
    result = CliRunner().invoke(fio_buffer.core.buffer, [
        'tests/data/points.geojson',
        outfile,
        '--distance', '100',
        '--driver', 'GeoJSON',
        '--buf-crs', 'EPSG:3857'
    ])
    assert result.exit_code == 0
    with fio.open('tests/data/points.geojson') as points, fio.open(outfile) as actual:
        for pnt, buf in zip(points, actual):

            # Reproject, buffer, reroject to create an expected geometry
            geom = transform_geom('EPSG:4326', 'EPSG:3857', pnt['geometry'])
            buf_geom = mapping(shape(geom).buffer(100))
            e_coords = transform_geom('EPSG:3857', 'EPSG:4326', buf_geom)['coordinates'][0]

            a_coords = buf['geometry']['coordinates'][0]
            for e_pair, a_pair in zip(e_coords, a_coords):
                # Multipolygons can creep in and mess up the test so just skip them
                if buf['geometry']['type'] == 'Polygon':
                    e_x, e_y = e_pair
                    a_x, a_y = a_pair
                    assert round(e_x, 3) == round(a_x, 3)
                    assert round(e_y, 3) == round(a_y, 3)


def test_buf_dst_crs(tmpdir):
    outfile = str(tmpdir.mkdir('out').join("buffered.geojson"))
    result = CliRunner().invoke(fio_buffer.core.buffer, [
        'tests/data/points.geojson',
        outfile,
        '--distance', '100',
        '--driver', 'GeoJSON',
        '--buf-crs', 'EPSG:3857',
        '--dst-crs', 'EPSG:900913'
    ])
    assert result.exit_code == 0
    with fio.open('tests/data/points.geojson') as points, fio.open(outfile) as actual:
        for pnt, buf in zip(points, actual):

            # Reproject, buffer, reroject to create an expected geometry
            geom = transform_geom('EPSG:4326', 'EPSG:3857', pnt['geometry'])
            buf_geom = mapping(shape(geom).buffer(100))
            e_coords = transform_geom('EPSG:3857', 'EPSG:900913', buf_geom)['coordinates'][0]

            a_coords = buf['geometry']['coordinates'][0]
            for e_pair, a_pair in zip(e_coords, a_coords):
                # Multipolygons can creep in and mess up the test so just skip them
                if buf['geometry']['type'] == 'Polygon':
                    e_x, e_y = e_pair
                    a_x, a_y = a_pair
                    assert round(e_x, 3) == round(a_x, 3)
                    assert round(e_y, 3) == round(a_y, 3)


def test_buffer_field(tmpdir):
    outfile = str(tmpdir.mkdir('out').join('buf-field'))
    result = CliRunner().invoke(fio_buffer.core.buffer, [
        'tests/data/points.geojson',
        outfile,
        '--distance', 'distance',
    ])
    assert result.exit_code == 0
    with fio.open('tests/data/points.geojson') as src, fio.open(outfile) as out:
        assert len(src) == len(out) and len(src) > 1
        for e, a in zip(src, out):
            assert e['properties'] == a['properties']
            e_coords = mapping(
                shape(e['geometry']).buffer(e['properties']['distance']))['coordinates'][0]
            a_coords = a['geometry']['coordinates'][0]
            for e_pair, a_pair in zip(e_coords, a_coords):
                e_x, e_y = e_pair
                a_x, a_y = a_pair
                assert round(e_x, 7) == round(a_x, 7)
                assert round(e_y, 7) == round(a_y, 7)


def test_register():
    # Make sure the plugin is actually registering
    assert 'buffer' in fiona.fio.main.main_group.commands
    result = CliRunner().invoke(
        fiona.fio.main.main_group, [
            'buffer',
            '--help'
        ]
    )
    assert result.exit_code == 0


def test_inherit_driver(tmpdir):
    outfile = str(tmpdir.mkdir('out').join("inherit-driver.geojson"))
    result = CliRunner().invoke(fio_buffer.core.buffer, [
        'tests/data/points.geojson',
        outfile,
        '--distance', '1',
    ])
    assert result.exit_code == 0
    with fio.open('tests/data/points.geojson') as src, fio.open(outfile) as out:
        assert src.driver == out.driver == 'GeoJSON'
        assert src.crs == out.crs == {'init': 'epsg:4326'}
        assert len(src) == len(out) and len(src) > 1


tests/test_processor.py
import copy

from shapely.geometry import mapping
from shapely.geometry import shape

import fio_buffer.core


feature = {
    'type': 'Feature',
    'properties': {'prop1': 1, 'prop2': None},
    'geometry': {
        'type': 'Polygon',
        'coordinates': [[[-1, -1], [-1, 1], [1, 1], [1, -1]]]
    }
}


def round7(val):
    return round(val, 7)


def test_just_buffer():
    args = {
        'feat': feature,
        'src_crs': None,
        'buf_crs': None,
        'dst_crs': None,
        'skip_failures': False,
        'buf_args': {'distance': 10},
    }

    expected = {
        'type': 'Feature',
        'properties': feature['properties'],
        'geometry': mapping(shape(feature['geometry']).buffer(**args['buf_args']))
    }

    actual = fio_buffer.core._processor(args)

    assert expected.keys() == actual.keys()
    assert expected['properties'] == actual['properties']

    for a_pair, e_pair in zip(
            expected['geometry']['coordinates'][0], actual['geometry']['coordinates'][0]):
        assert list(map(round7, a_pair)) == list(map(round7, e_pair))


def test_buffer_field():
    args = {
        'feat': feature,
        'src_crs': None,
        'buf_crs': None,
        'dst_crs': None,
        'skip_failures': False,
        'buf_args': {'distance': 'prop1'},
    }

    expected = {
        'type': 'Feature',
        'properties': feature['properties'],
        'geometry': mapping(shape(feature['geometry']).buffer(distance=1))
    }

    actual = fio_buffer.core._processor(args)

    assert expected.keys() == actual.keys()
    assert expected['properties'] == actual['properties']

    for a_pair, e_pair in zip(
            expected['geometry']['coordinates'][0], actual['geometry']['coordinates'][0]):
        assert list(map(round7, a_pair)) == list(map(round7, e_pair))


def test_buffer_field_no_val():
    # Buffer by a field where the distance is None
    args = {
        'feat': feature,
        'src_crs': None,
        'buf_crs': None,
        'dst_crs': None,
        'skip_failures': False,
        'buf_args': {'distance': 'prop2'},
    }

    expected = copy.deepcopy(feature)
    actual = fio_buffer.core._processor(args)

    assert expected.keys() == actual.keys()
    assert expected['properties'] == actual['properties']

    for a_pair, e_pair in zip(
            expected['geometry']['coordinates'][0], actual['geometry']['coordinates'][0]):
        assert list(map(round7, a_pair)) == list(map(round7, e_pair))

tests/test_exceptions.py
from click.testing import CliRunner

import fio_buffer.core


def test_dst_crs_no_buf_crs():
    result = CliRunner().invoke(fio_buffer.core.buffer, [
        'tests/data/points.geojson',
        'should-not-be-written',
        '--driver', 'GeoJSON',
        '--dist', '1',
        '--dst-crs', 'EPSG:3857'
    ])
    assert result.exit_code != 0


fio_buffer/__init__.py
"""
A Fiona CLI plugin for buffering geometries.
"""


__version__ = '0.1.1'
__author__ = 'Kevin Wurster'
__email__ = 'wursterk@gmail.com'
__source__ = 'https://github.com/geowurster/fio-buffer'
__license__ = '''
New BSD License

Copyright (c) 2015, Kevin D. Wurster
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* The names of fio-buffer or its contributors may not be used to endorse or
  promote products derived from this software without specific prior written
  permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
'''


fio_buffer/core.py


"""
Core components for fio-buffer
"""


import copy
import logging
from multiprocessing import cpu_count, Pool

import click
import fiona as fio
from fiona.transform import transform_geom
from shapely.geometry import CAP_STYLE
from shapely.geometry import JOIN_STYLE
from shapely.geometry import mapping
from shapely.geometry import shape

from . import __version__


logging.basicConfig()
logger = logging.getLogger('fio-buffer')


def _cb_cap_style(ctx, param, value):

    """
    Click callback to transform `--cap-style` to an `int`.
    """

    return getattr(CAP_STYLE, value)


def _cb_join_style(ctx, param, value):

    """
    Click callback to transform `--join-style` to an `int`.
    """

    return getattr(JOIN_STYLE, value)


def _cb_res(ctx, param, value):

    """
    Click callback to ensure `--res` is `>= 0`.
    """

    if value < 0:
        raise click.BadParameter("must be a positive value")

    return value


def _cb_dist(ctx, param, value):

    """
    Click callback to ensure `--distance` can be either a float or a field name.
    """

    try:
        return float(value)
    except ValueError:
        return value


def _processor(args):

    """
    Process a single feature.

    Parameters
    ----------
    args : dict
        feat : dict
            A GeoJSON feature to process.
        src_crs : str or dict
            The geometry's CRS.
        buf_crs : str or dict
            Apply buffer after reprojecting to this CRS.
        dst_crs : str or dict
            Reproject buffered geometry to this CRS before returning.
        skip_failures : bool
            If True then Exceptions don't stop processing.
        buf_args : dict
            Keyword arguments for the buffer operation.

    Returns
    -------
    dict
        GeoJSON feature with updated geometry.
    """

    feat = args['feat']
    src_crs = args['src_crs']
    buf_crs = args['buf_crs']
    dst_crs = args['dst_crs']
    skip_failures = args['skip_failures']
    buf_args = args['buf_args']

    # Support buffering by a field's value
    if not isinstance(buf_args['distance'], (float, int)):
        field_val = feat['properties'][buf_args['distance']]

        # Buffering according to a field but field is None so just return the feature
        if field_val is None:
            return feat
        else:
            buf_args['distance'] = field_val

    try:
        # src_crs -> buf_crs
        reprojected = transform_geom(
            src_crs, buf_crs, feat['geometry'],
            antimeridian_cutting=True
        )

        # buffering operation
        buffered = shape(reprojected).buffer(**buf_args)

        # buf_crs -> dst_crs
        feat['geometry'] = transform_geom(
            buf_crs, dst_crs, mapping(buffered),
            antimeridian_cutting=True
        )

        return feat

    except Exception:
        logger.exception("Feature with ID %s failed during buffering", feat.get('id'))
        if not skip_failures:
            raise


@click.command(short_help="Buffer geometries on all sides by a fixed distance.")
@click.version_option(prog_name='fio-buffer', version=__version__)
@click.argument('infile', required=True)
@click.argument('outfile', required=True)
@click.option(
    '-f', '--format', '--driver', metavar='NAME',
    help="Output driver name.  Derived from the input datasource if not given."
)
@click.option(
    '--cap-style', type=click.Choice(['flat', 'round', 'square']),
    default='round', show_default=True,
    callback=_cb_cap_style, help="Where geometries terminate, use this style."
)
@click.option(
    '--join-style', type=click.Choice(['round', 'mitre', 'bevel']),
    default='round', show_default=True,
    callback=_cb_join_style, help="Where geometries touch, use this style."
)
@click.option(
    '--res', type=click.INT, callback=_cb_res, default=16, show_default=True,
    help="Resolution of the buffer around each vertex of the geometry."
)
@click.option(
    '--mitre-limit', type=click.FLOAT, default=5.0, show_default=True,
    help="When using a mitre join, limit the maximum length of the join corner according to "
         "this ratio."
)
@click.option(
    '--distance', metavar='FLOAT|FIELD', required=True, callback=_cb_dist,
    help="Buffer distance or field containing distance values.  Units match --buf-crs.  "
         "When buffering with a field, feature's with a null value are unaltered."
)
@click.option(
    '--src-crs', help="Specify CRS for input data.  Not needed if set in input file."
)
@click.option(
    '--buf-crs', help="Perform buffer operations in a different CRS. [default: --src-crs]"
)
@click.option(
    '--dst-crs', help="Reproject geometries to a different CRS before writing.  Must be "
                      "combined with --buf-crs. [default: --src-crs]"
)
@click.option(
    '--geom-type', 'output_geom_type', default='MultiPolygon',
    metavar='GEOMTYPE', show_default=True,
    help="Output layer's geometry type."
)
@click.option(
    '--skip-failures', is_flag=True,
    help="Skip geometries that fail somewhere in the processing pipeline."
)
@click.option(
    '--jobs', type=click.IntRange(1, cpu_count()), default=1,
    metavar="CORES", show_default=True,
    help="Process geometries in parallel across N cores.  Feature ID's and order are not "
         "preserved if more that 1 cores are used."
)
@click.pass_context
def buffer(ctx, infile, outfile, driver, cap_style, join_style, res, mitre_limit,
           distance, src_crs, buf_crs, dst_crs, output_geom_type, skip_failures, jobs):

    """
    Geometries can be dilated with a positive distance, eroded with a negative
    distance, and in some cases cleaned or repaired with a distance of 0.

    \b
    Examples
    --------

    Default settings - buffer geometries in the input CRS:

    \b
        $ fio buffer in.geojson out.geojson --distance 10

    Dynamically buffer geometries by a distance stored in the field 'magnitude'
    and write as GeoJSON:

    \b
        $ fio buffer \\
            in.shp \\
            out.geojson \\
            --driver GeoJSON \\
            --distance magnitude

    Read geometries from one CRS, buffer in another, and then write to a third:

    \b
        $ fio buffer in.shp out.shp \\
            --distance 10 \\
            --buf-crs EPSG:3857 \\
            --dst-crs EPSG:32618

    Control cap style, mitre limit, segment resolution, and join style:

    \b
        $ fio buffer in.geojson out.geojson \\
            --distance 0.1 \\
            --res 5 \\
            --cap-style flat \\
            --join-style mitre \\
            --mitre-limit 0.1\\
    """

    if dst_crs and not buf_crs:
        raise click.ClickException("Must specify --buf-crs when using --dst-crs.")

    # fio has a -v flag so just use that to set the logging level
    # Extra checks are so this plugin doesn't just completely crash due
    # to upstream changes.
    if isinstance(getattr(ctx, 'obj'), dict):
        logger.setLevel(ctx.obj.get('verbosity', 1))

    with fio.open(infile) as src:

        logger.debug("Resolving CRS fall backs")

        src_crs = src_crs or src.crs
        buf_crs = buf_crs or src_crs
        dst_crs = dst_crs or src_crs

        if not src_crs:
            raise click.ClickException(
                "CRS is not set in input file.  Use --src-crs to specify.")

        logger.debug("src_crs=%s", src_crs)
        logger.debug("buf_crs=%s", buf_crs)
        logger.debug("dst_crs=%s", dst_crs)

        meta = copy.deepcopy(src.meta)
        meta.update(
            driver=driver or src.driver,
            crs=dst_crs
        )
        if output_geom_type:
            meta['schema'].update(geometry=output_geom_type)

        logger.debug("Creating output file %s", outfile)
        logger.debug("Meta=%s", meta)

        with fio.open(outfile, 'w', **meta) as dst:

            # Keyword arguments for `<Geometry>.buffer()`
            buf_args = {
                'distance': distance,
                'resolution': res,
                'cap_style': cap_style,
                'join_style': join_style,
                'mitre_limit': mitre_limit
            }

            # A generator that produces the arguments required for `_processor()`
            task_generator = (
                {
                    'feat': feat,
                    'src_crs': src_crs,
                    'buf_crs': buf_crs,
                    'dst_crs': dst_crs,
                    'skip_failures': skip_failures,
                    'buf_args': buf_args
                } for feat in src)

            logger.debug("Starting processing on %s cores", jobs)
            for o_feat in Pool(jobs).imap_unordered(_processor, task_generator):
                if o_feat is not None:
                    try:
                        dst.write(o_feat)
                    except Exception:
                        logger.exception(
                            "Feature with ID %s failed during write", o_feat.get('id'))
                        if not skip_failures:
                            raise

            logger.debug("Finished processing.")
